GoogleForCreators/web-stories-wp
# Changed white-space handling in text elements

## Summary

This changes how white-space is handled in text elements, as Firefox is unable to make justified text, if white-space is set tp `pre-wrap`. Setting it to `pre-line` does (almost) the same thing while avoiding the bug in Firefox completely.

The main difference between `pre-wrap` and `pre-line` are:

- Multiple white-space characters in a row are collapsed (unless non-breaking).
- Leading white-space is removed (unless non-breaking).

Potential consequences:

- If any text sets use either of the above, they would render wrong. However, both searching as well as manual checking seem to indicate, that no text sets do this.
- If any templates (story and/or page) use either of the above, they would render wrong. It was a lot harder to check these, but it does not appear that any of them use either of the above.
- If a user has a story with multiple spaces or leading space in a line, it could potentially render wrong. However, because our text editor converts all spaces to non-breaking spaces and all tabs (they could be pasted in) to four non-breaking spaces, this is never actually a problem. Multiple spaces manually entered in the editor still render correctly!

## User-facing changes

Various browsers with various text alignment, before and after:

| Browser | Before | After |
|-|-|-|
| Chrome | ![chrome_before](https://user-images.githubusercontent.com/637548/163247462-04ac01ff-5534-4d73-8077-6b2f166d94cc.png) | ![chrome_after](https://user-images.githubusercontent.com/637548/163247476-2d6b114e-16cd-41b7-b274-6b538870b143.png) |
| Safari | ![safari_before](https://user-images.githubusercontent.com/637548/163247521-e6029aba-de1c-4e08-a815-bcefc71e7319.png) | ![safari_after](https://user-images.githubusercontent.com/637548/163247540-4c29b0b7-5e3a-461d-90fb-e01661cf14f9.png) |
| Firefox | ![firefox_before](https://user-images.githubusercontent.com/637548/163247553-485cad6a-e9e8-4030-bd9b-42615d5f591d.png) | ![firefox_after](https://user-images.githubusercontent.com/637548/163247569-8b9751f8-fc65-4488-90e4-9dae21efb727.png) |

Note however how different browsers treat multiple spaces differently! Between the red letters, there are two (non-breaking) spaces, between the other words only one space. The differences are not completely logical to me, but definitely there:
![pre-line-rendering-differences](https://user-images.githubusercontent.com/637548/163603266-161cfd1b-b206-42ed-8062-5cd97365368d.jpg)

## Testing Instructions

This PR can be tested by following these steps:

1.  Pay special attention to Firefox, but test in other browsers too
2. Add a text field and make sure it has multiple lines due to wrapping, not due to manual line-breaks
3. Adjust text alignment to justified
4. Observe that text actually renders justified in both editor and final story
5. Play with multiple spaces, newlines, and leading spaces after a newline and observe, that all white-space renders correctly in both editor and final story

Known bug with Firefox + Justified Text + Edit-mode:

1. Add a text field in Firefox
2. Add multiple lines of text
3. Set text alignment to justified
4. Click inside text to enter edit mode
5. Observe that text is not longer justified but left-aligned. This is a known and acceptable bug and it is better than the alternative.
6. Exit editmode and observe that text once again appears justified
7. Preview the story
8. Observe that text is correctly justified

## Checklist




---

Fixes #10542
